### PR TITLE
Don't recreate iterator for each series on each timestep when evaluating a query with `timestamp()`

### PR DESCRIPTION
**Pull request by charleskorn**
While investigating an issue with queries that use the `timestamp()` function with Mimir's chunks streaming (https://github.com/grafana/mimir/pull/5370), we discovered that the current special case handling for this function creates a new series iterator for each series for each time step in the query, as well as a new `MemoizedSeriesIterator` for each time step.

This turns out to be quite inefficient. Changing the special case for `timestamp()` to create and retain one iterator per series for the whole query has the following impact on the benchmark added in this PR:

```
goos: darwin
goarch: arm64
pkg: github.com/prometheus/prometheus/promql
                                                   │  before.txt  │              after.txt              │
                                                   │    sec/op    │    sec/op     vs base               │
RangeQuery/expr=timestamp(a_one),steps=1-10           10.74µ ± 1%   11.84µ ±  7%  +10.19% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_one),steps=10-10          25.56µ ± 1%   12.54µ ±  7%  -50.95% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_one),steps=100-10        271.63µ ± 1%   36.07µ ±  6%  -86.72% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_one),steps=1000-10       8798.2µ ± 1%   221.0µ ± 11%  -97.49% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_ten),steps=1-10           48.00µ ± 1%   36.20µ ±  9%  -24.58% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_ten),steps=10-10         187.21µ ± 1%   56.89µ ± 10%  -69.61% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_ten),steps=100-10        2600.1µ ± 1%   290.3µ ±  8%  -88.84% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_ten),steps=1000-10       89.292m ± 2%   2.562m ±  7%  -97.13% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_hundred),steps=1-10       416.8µ ± 1%   322.0µ ± 20%  -22.74% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_hundred),steps=10-10     1816.6µ ± 1%   562.5µ ±  7%  -69.04% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_hundred),steps=100-10    26.112m ± 1%   2.542m ± 20%  -90.26% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_hundred),steps=1000-10   896.10m ± 1%   22.67m ±  1%  -97.47% (p=0.002 n=6)
geomean                                               1.304m        227.2µ        -82.58%

                                                   │  before.txt   │              after.txt              │
                                                   │     B/op      │     B/op      vs base               │
RangeQuery/expr=timestamp(a_one),steps=1-10           7.506Ki ± 0%   7.013Ki ± 0%   -6.57% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_one),steps=10-10         12.782Ki ± 0%   7.787Ki ± 0%  -39.08% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_one),steps=100-10         68.79Ki ± 0%   15.76Ki ± 0%  -77.09% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_one),steps=1000-10       994.03Ki ± 0%   96.04Ki ± 0%  -90.34% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_ten),steps=1-10           21.54Ki ± 0%   22.25Ki ± 0%   +3.26% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_ten),steps=10-10          47.25Ki ± 0%   28.25Ki ± 0%  -40.21% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_ten),steps=100-10        339.33Ki ± 0%   89.54Ki ± 0%  -73.61% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_ten),steps=1000-10       6996.4Ki ± 0%   703.0Ki ± 0%  -89.95% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_hundred),steps=1-10       169.1Ki ± 0%   181.8Ki ± 0%   +7.57% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_hundred),steps=10-10      398.3Ki ± 0%   239.4Ki ± 0%  -39.90% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_hundred),steps=100-10    3044.3Ki ± 0%   826.3Ki ± 0%  -72.86% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_hundred),steps=1000-10   65.363Mi ± 0%   6.540Mi ± 0%  -89.99% (p=0.002 n=6)
geomean                                               278.5Ki        98.03Ki       -64.80%

                                                   │  before.txt   │             after.txt              │
                                                   │   allocs/op   │  allocs/op   vs base               │
RangeQuery/expr=timestamp(a_one),steps=1-10             152.0 ± 0%    147.0 ± 0%   -3.29% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_one),steps=10-10            233.0 ± 0%    174.0 ± 0%  -25.32% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_one),steps=100-10          1192.0 ± 0%    447.0 ± 0%  -62.50% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_one),steps=1000-10        19.886k ± 0%   3.167k ± 0%  -84.07% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_ten),steps=1-10             290.0 ± 0%    285.0 ± 0%   -1.72% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_ten),steps=10-10            624.0 ± 0%    322.0 ± 0%  -48.40% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_ten),steps=100-10          5686.0 ± 0%    710.0 ± 0%  -87.51% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_ten),steps=1000-10       153.082k ± 0%   4.561k ± 0%  -97.02% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_hundred),steps=1-10        1.584k ± 0%   1.579k ± 0%   -0.32% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_hundred),steps=10-10       4.404k ± 0%   1.672k ± 0%  -62.03% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_hundred),steps=100-10     50.081k ± 0%   2.801k ± 0%  -94.41% (p=0.002 n=6)
RangeQuery/expr=timestamp(a_hundred),steps=1000-10   1480.41k ± 0%   13.78k ± 0%  -99.07% (p=0.002 n=6)
geomean                                                4.497k        1.008k       -77.59%

```

Note to reviewers: I'd suggest reviewing each commit separately, as they incrementally build up the changes required.

### CODE REVIEW

**Review by pracucci — APPROVED** [review]
I reviewed these changes in `grafana/mimir-prometheus` (where we stage experimental changes before proposing them upstream) and FWIW LGTM. FYI these changes are already running in production at Grafana Labs.

File: promql/bench_test.go [hunk 1]
```diff
@@ -186,6 +186,10 @@ func rangeQueryCases() []benchCase {
 			expr:  "count({__name__!=\"\",l=\"\"})",
 			steps: 1,
 		},
+		// timestamp() function
```
Comment (bboreham) [inline promql/bench_test.go:189]: comment doesn't add much

File: promql/engine_test.go [hunk 2]
```diff
@@ -1977,6 +1977,100 @@ func TestSubquerySelector(t *testing.T) {
...
+	start := time.Unix(0, 0)
+	end := time.Unix(61, 0)
+	interval := time.Second
+
+	expectedResult := Matrix{
+		Series{
+			Floats: []FPoint{
+				{F: 0, T: 0},
+				{F: 0, T: 1_000},
+				{F: 0, T: 2_000},
```
Comment (bboreham) [inline promql/engine_test.go:2001]: Can this be done with a loop?

File: promql/engine.go [hunk 3]
```diff
@@ -1387,15 +1387,7 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
...
-				return ev.rangeEval(nil, func(v []parser.Value, _ [][]EvalSeriesHelper, enh *EvalNodeHelper) (Vector, storage.Warnings) {
-					if vs.Timestamp != nil {
-						// This is a special case only for "timestamp" since the offset
-						// needs to be adjusted for every point.
-						vs.Offset = time.Duration(enh.Ts-*vs.Timestamp) * time.Millisecond
-					}
-					val, ws := ev.vectorSelector(vs, enh.Ts)
-					return call([]parser.Value{val}, e.Args, enh), ws
-				})
+				return ev.evalTimestampFunctionOverVectorSelector(vs, call, e)
```
Comment (bboreham) [inline promql/engine.go:1390]: I wondered if this should be called `rangeEvalTimestampFunctionOverVectorSelector`

File: promql/engine.go [hunk 4]
```diff
@@ -1833,38 +1825,47 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
...
 
-			ev.currentSamples++
-			ev.samplesStats.IncrementSamplesAtTimestamp(ts, 1)
-			if ev.currentSamples > ev.maxSamples {
-				ev.error(ErrTooManySamples(env))
-			}
+	return ev.rangeEval(nil, func(v []parser.Value, _ [][]EvalSeriesHelper, enh *EvalNodeHelper) (Vector, storage.Warnings) {
+		if vs.Timestamp != nil {
+			// This is a special case only for "timestamp" since the offset
+			// needs to be adjusted for every point.
```
Comment (bboreham) [inline promql/engine.go:1843]: Could be rephrased, inside a function specifically for `timestamp()` .
Maybe it should be part of the doc comment for the function.

**Review by bboreham — COMMENTED** [review]
Generally very good, thanks.

If I understood, some part of the change is necessary to permit streaming processing of chunks.  Should there be a comment somewhere to that effect, in case someone has a cool idea to change it and break that property?

File: promql/bench_test.go (hunk 1 above)
Comment (charleskorn) [reply promql/bench_test.go:189]: I agree - I was following the existing pattern in this test of including a comment before each group of test cases. Happy to remove it if you'd prefer.

File: promql/engine.go (hunk 3 above)
Comment (charleskorn) [reply promql/engine.go]: Makes sense to me, renamed in 28f8773

File: promql/engine.go (hunk 4 above)
Comment (charleskorn) [reply promql/engine.go]: I've added further clarification in cd3e0b1 - does this cover what you had in mind?

File: promql/engine_test.go (hunk 2 above)
Comment (charleskorn) [reply promql/engine_test.go:2001]: Yes - resolved in 0a0f993

File: promql/bench_test.go (hunk 1 above)
Comment (bboreham) [reply promql/bench_test.go:189]: Under that pattern it could be something like "Functions which have special handling inside 'eval'".

**Review by bboreham — APPROVED** [review]
lgtm

File: promql/bench_test.go (hunk 1 above)
Comment (charleskorn) [reply promql/bench_test.go]: Makes sense, changed in d396282

**Pull request merged**